### PR TITLE
Edge mixer template. Stackdriver Adapter: Add support to write to Context Graph

### PR DESCRIPTION
**Pull request by jeffmendoza**
This includes #6408, so I'll close that one.

Add Mixer template for edge instances. These can be used by adapters to construct graphs of the Istio mesh.

Add support to Stackdriver adapter to write to the Stackdriver Context Graph API.

The Context Graph client is not available upstream yet. So vendor it inside the Stackdriver adapter until it is available.

### CODE REVIEW

File: mixer/adapter/stackdriver/contextgraph/contextgraph.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		projectID:      b.projectID,
+		meshUID:        fmt.Sprintf("%s/%s/%s", b.projectID, b.zone, b.cluster),
+		zone:           b.zone,
+		cluster:        b.cluster,
+		entityCache:    newEntityCache(env.Logger()),
+		edgeCache:      newEdgeCache(env.Logger()),
+		traffics:       make(chan trafficAssertion),
+		entitiesToSend: make([]entity, 0),
+		edgesToSend:    make([]edge, 0),
+		sendTick:       time.NewTicker(30 * time.Second),
```
Comment (douglas-reid) [inline mixer/adapter/stackdriver/contextgraph/contextgraph.go]: should `30s` be a const, and/or configurable?

File: mixer/adapter/stackdriver/contextgraph/contextgraph.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import (
+	"context"
+	"fmt"
+	"time"
+
+	contextgraph "cloud.google.com/go/contextgraph/apiv1alpha1"
+	gax "github.com/googleapis/gax-go"
+
+	"google.golang.org/api/option"
+	contextgraphpb "google.golang.org/genproto/googleapis/cloud/contextgraph/v1alpha1"
```
Comment (douglas-reid) [inline mixer/adapter/stackdriver/contextgraph/contextgraph.go]: style nit: we should separate out the istio imports section from the third-party. these two lines belong with the others.

File: mixer/adapter/stackdriver/contextgraph/contextgraph_test.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	if b.projectID != c.ProjectId {
+		t.Error("Expected builder projectID to be set, wasn't")
+	}
+
+	if mg.GMCalled == false {
+		t.Error("Expected GenerateMetadata to be called, wasn't")
+	}
+}
+
+type mockLogger struct{}
```
[the remark this answers is not in the document]
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/contextgraph_test.go]: Yes, let me try this out.

File: mixer/adapter/stackdriver/contextgraph/send.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+					h.edgeCache.Invalidate(entity.fullName)
+				}
+			}
+			for _, edge := range edges {
+				if h.edgeCache.AssertAndCheck(edge, epoch) {
+					h.edgesToSend = append(h.edgesToSend, edge)
+				}
+			}
+		case t := <-h.sendTick.C:
+			if t.After(lastFlush.Add(9 * time.Minute)) {
```
[the remark this answers is not in the document]
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/send.go]: No, it is a detail of the ContextGraph API, and wouldn't be good for the user to configure.

File: mixer/adapter/stackdriver/contextgraph/contextgraph.go (hunk 1 above)
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/contextgraph.go]: I don't think it should be user configurable, but maybe. The ContextGraph API has a very low quota, and expects us to batch up asserts.

File: mixer/adapter/stackdriver/contextgraph/contextgraph.go (hunk 2 above)
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/contextgraph.go]: Will do.

File: mixer/adapter/stackdriver/contextgraph/cache.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import "istio.io/istio/mixer/pkg/adapter"
+
+type cacheStatus struct {
+	lastSeen int
+	lastSent int
+}
+
+// entityCache tracks the entities we've already seen.
+// It is not thread-safe.
+type entityCache struct {
```
[the remark this answers is not in the document]
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/cache.go]: @quentinmit Can you take a look? Thanks

File: mixer/adapter/stackdriver/contextgraph/contextgraph.go (hunk 2 above)
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/contextgraph.go]: Done

File: mixer/adapter/stackdriver/contextgraph/contextgraph_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeffmendoza) [reply mixer/adapter/stackdriver/contextgraph/contextgraph_test.go]: Done

File: mixer/adapter/stackdriver/contextgraph/cache.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		ec.cache[k] = e
+		result = append(result, k)
+	}
+	ec.lastFlush = epoch
+	return result
+}
+
+// edgeCache tracks the edges we've already seen.
+// It is not thread-safe.
+type edgeCache struct {
```
[the remark this answers is not in the document]
Comment (quentinmit) [reply mixer/adapter/stackdriver/contextgraph/cache.go]: Not until generics land in Go 2 ;)

File: mixer/adapter/stackdriver/contextgraph/send.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (quentinmit) [reply mixer/adapter/stackdriver/contextgraph/send.go]: Context API requires a refresh every 10m. This is something clients are just supposed to hardcode.

**Comment by douglas-reid** [comment]
We need this PR in ASAP to make the 1.0 release cut. We can make small adjustments to the code moving forward, but I see no major blockers in my review.

/lgtm
/approve
/retest

**Comment by jeffmendoza** [comment]
@douglas-reid Yay, all tests pass.

**Comment by jeffmendoza** [comment]
Docker Hub!!
```
 error pulling image configuration: Get https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/21/219e59ae4605f1a61d4f33dbc6dc46f170a8be109b3ba59cba317b7633a87b9b/data?verify=1530379542-2YTxCF70uquWqfVJSgIQqGuQg98%3D: dial tcp: lookup production.cloudflare.docker.com on 10.47.240.10:53: server misbehaving
```

**Pull request merged**